File: app/src/main/java/sgtmelon/scriptum/app/control/input/InputTextWatcher.java
package sgtmelon.scriptum.app.control.input;

import android.text.Editable;
import android.text.TextWatcher;
import android.widget.EditText;

import androidx.annotation.NonNull;
import sgtmelon.scriptum.app.model.item.CursorItem;
import sgtmelon.scriptum.office.annot.def.InputDef;

/**
 * Контроллер ввода текста для {@link InputControl}
 */
public final class InputTextWatcher implements TextWatcher {

    private final EditText view;
    private final int tag;

    private final Result result;
    private final InputIntf inputIntf;

    private String textFrom = "";
    private int cursorFrom = 0;

    // TODO nullable view

    public InputTextWatcher(EditText view, @InputDef int tag,
                            @NonNull Result result, @NonNull InputIntf inputIntf) {
        this.view = view;
        this.tag = tag;

        this.result = result;
        this.inputIntf = inputIntf;
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        textFrom = s.toString();
        cursorFrom = view.getSelectionEnd();
    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
        final String textTo = s.toString();
        final int cursorTo = view.getSelectionEnd();

        if (textFrom.equals(textTo)) return;

        final CursorItem cursorItem = new CursorItem(cursorFrom, cursorTo);

        switch (tag) {
            case InputDef.name:
                inputIntf.onNameChange(textFrom, textTo, cursorItem);
                break;
            case InputDef.text:
                inputIntf.onTextChange(textFrom, textTo, cursorItem);
                break;
        }

        textFrom = textTo;
        cursorFrom = cursorTo;

        result.onTextResult();
    }

    @Override
    public void afterTextChanged(Editable s) {

    }

    public interface Result {
        void onTextResult();
    }

}

File: app/src/main/java/sgtmelon/scriptum/app/control/input/InputTextWatcher.kt
package sgtmelon.scriptum.app.control.input

import android.text.Editable
import android.text.TextWatcher
import android.widget.EditText
import sgtmelon.scriptum.app.model.item.CursorItem

/**
 * Контроллер ввода текста для [InputControl]
 */
class InputTextWatcher(private val view: EditText?,
                       @param:InputDef private val tag: Int,
                       private val result: Result,
                       private val inputCallback: InputCallback
) : TextWatcher {

    private var textFrom = ""
    private var cursorFrom = 0

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
        textFrom = s.toString()
        cursorFrom = view?.selectionEnd ?: 0
    }

    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
        val textTo = s.toString()
        val cursorTo = view?.selectionEnd ?: 0

        if (textFrom == textTo) return

        val cursorItem = CursorItem(cursorFrom, cursorTo)

        //TODO вынести в интерфейс результата
        when (tag) {
            InputDef.name -> inputCallback.onNameChange(textFrom, textTo, cursorItem)
            InputDef.text -> inputCallback.onTextChange(textFrom, textTo, cursorItem)
        }

        textFrom = textTo
        cursorFrom = cursorTo

        result.onInputTextChangeResult()
    }

    override fun afterTextChanged(s: Editable) {}

    interface Result {
        fun onInputTextChangeResult()
    }

}